Add next-gen beta documentation (#1565)
slackapi/bolt-js: docs/_future/custom_functions.md

  try {
    await client.chat.postMessage({
      channel,
      text: `${salutation}, <@${recipient}>! :wave: Someone sent the following greeting: \n\n>${message}`,
    });
    complete();
  } catch (err) {
    // Complete function with an error
    await complete({ error: `There was an issue: ${err}` });
    throw (err);
  }
};

// Let's create a new Slack Function with helloWorld as its handler
const helloWorldFunc = new SlackFunction(SampleFunctionDefinition.id, helloWorld);

module.exports = { helloWorldFunc };
```

Once this file has been created, the function handler can be registered in an `index.js` file within the `listeners/functions` directory:
```js
const { helloWorldFunc } = require('./hello-world');

// Register a complete function
module.exports.register = (app) => {
  app.function(helloWorldFunc);
  // Register another function here
};
```

In order to make sure this handler is triggered, make sure the Function listeners are registered in your `listeners/index.js` file:
```js
const functions = require('./functions');

module.exports.registerListeners = (app) => {
  functions.register(app);
};
```
#### 4. Add function to app through workflow or manifest

> To actually call the defined function, `SampleFunctionDefinition`, don't forget to add your function to a workflow! When you're finished defining and implementing your functions, the next step is to add them to [Workflows](/bolt-js/future/workflows). Once added as a step in a Workflow, your Function will run when that Workflow is invoked by a [Trigger](/bolt-js/future/triggers).

Reviewer: @srajiang I omitted the distribution of functions since it required deployment - let me know if you feel that we should add it back in based on what we decide to include in the deploy guide! (cc: @e-zim)

If you haven't done so already, create a file for your function definition to live in and name it something that makes sense for your function. In the [Bolt for JavaScript Starter Template](https://github.com/slack-samples/bolt-js-starter-template/blob/future/manifest/function/sample-function.js), we named this file `sample-function.js`

#### 2. Add function listener and handler(s)
Once your function has been defined, you'll need to register a function listener to trigger the function's functionality as well as any other related events.

To do this, create a file in `listeners/functions` directory for your function listener (in this case, the file will be called `hello-world.js`). This file will use the function definition you created earlier and also implement any needed additional logic&mdash;in this case, it'll send a message to greet someone based on what is being passed in as inputs. You can see the full completed `hello-world.js` function listener [here](https://github.com/slack-samples/bolt-js-starter-template/blob/future/listeners/functions/hello-world.js).

```js
// listeners/functions/hello-world.js

> // For more information about functions: https://api.slack.com/future/functions

Reviewer 2: ```suggestion
// For more information about functions: https://api.slack.com/future/functions
```

I'd remove this for now, since I don't want devs getting lost in deno samples and code snippets

---
title: Custom functions
order: 5
slug: custom-functions
lang: en
layout: tutorial
permalink: /future/custom-functions
---
# Custom functions <span class="label-beta">BETA</span>

<div class="section-content">
On the next-generation platform, you can build **custom functions**, reusable building blocks of automation that are deployed to our infrastructure and accept inputs, perform some calculations, and provide outputs. Functions can be used as steps in [Workflows](/bolt-js/future/workflows)&mdash;and Workflows are invoked by [Triggers](/bolt-js/future/triggers).

To create a function, we need to do the following: 
- [define the function](#define) in the Manifest;
- [implement the function](#implement) in its respective source file;
</div>

---

### Defining a function {#define}
Functions are defined in your app via the `DefineFunction` method, which is part of the SDK that gets included with every newly created project. These function definitions are stored under the `manifest/functions/` directory.

> Let's go ahead and create a new function definition file under `manfiest/functions` directory. Name it something related to what the function does. In our [Bolt for JavaScript Starter Template](https://github.com/slack-samples/bolt-js-starter-template/blob/future/manifest/function/sample-function.js), we name the file `sample-function.js`. Let's take a peak at it:

Reviewer: ```suggestion
Let's go ahead and create a new function definition file under `manfiest/functions` directory. Name it something related to what the function does. In our [Bolt for JavaScript Starter Template](https://github.com/slack-samples/bolt-js-starter-template/blob/future/manifest/function/sample-function.js), we name the file `sample-function.js`. Let's take a peek at it:
```

Functions are similar in philosophy to Unix system commands: they should be minimalist, modular, and reusable. Expect the output of one function to eventually become the input of another, with no other frame of reference.

Once your function is defined in its own file in `manifest/functions`, the next steps are to add a function listener, register the listener, and implement the function in a workflow.

---

### Implementing a Function Listener {#implement}

Implement functions in just a few steps:

#### 1. Create the function definition file in the `manifest/functions` directory
If you haven't done so already, create a file for your function definition to live in and name it something that makes sense for your function. In the [Bolt for JavaScript Starter Template](https://github.com/slack-samples/bolt-js-starter-template/blob/future/manifest/function/sample-function.js), we named this file `sample-function.js`

#### 2. Add function listener and handler(s)
Once your function has been defined, you'll need to register a function listener to trigger the function's functionality as well as any other related events.

> To do this, create a file in `listeners/functions` directory for your function listener (in this case, the file will be called `hello-world.js`). This file will use the function definition you created earlier and also implement any needed additional logic&mdash;in this case, it'll send a message to greet someone based on what is being passed in as inputs. 

Reviewer: ```suggestion
To do this, create a file in `listeners/functions` directory for your function listener (in this case, the file will be called `hello-world.js`). This file will use the function definition you created earlier and also implement any needed additional logic&mdash;in this case, it'll send a message to greet someone based on what is being passed in as inputs. You can see the full completed `hello-world.js` function listener [here](https://github.com/slack-samples/bolt-js-starter-template/blob/future/listeners/functions/hello-world.js).
```

It looks like we have a link to the full `hello-world.js` file below on line 162. Since it's a bit lengthy, what if we removed the code snippet below and updated this paragraph with this to link out to the snippet?

```js
// listeners/functions/hello-world.js
// For more information about functions: https://api.slack.com/future/functions
const { SlackFunction } = require('@slack/bolt');

// Get our Function Definition from the manifest!
const { SampleFunctionDefinition } = require('../../manifest/functions/sample-function');

// Here is the work we want to do!
const helloWorld = async ({ event, client, complete }) => {
  const { recipient, channel, message } = event.inputs;
  const salutations = ['Hello', 'Hi', 'Howdy', 'Hola', 'Salut'];
  const salutation = salutations[Math.floor(Math.random() * salutations.length)];
  try {
    await client.chat.postMessage({
      channel,
      text: `${salutation}, <@${recipient}>! :wave: Someone sent the following greeting: \n\n>${message}`,
    });
    complete();
  } catch (err) {
    // Complete function with an error
    await complete({ error: `There was an issue: ${err}` });
    throw (err);
  }
};

// Let's create a new Slack Function with helloWorld as its handler
const helloWorldFunc = new SlackFunction(SampleFunctionDefinition.id, helloWorld);

module.exports = { helloWorldFunc };
```

> The above file declares a function handler, `helloWorld`, that takes in inputs from the `event`, which is the payload received when your function is being executed. It executes logic within the handler to send a message with a random greeting to the specified recipient in the desired channel. Then, a new `SlackFunction` instance is declared that actually links the `helloWorld` handler to the function being run through its function ID (`SampleFunctionDefinition`):

Reviewer: ```suggestion
The `hello-world.js` file declares a function handler, `helloWorld`, that takes in inputs from the `event`, which is the payload received when your function is being executed. It executes logic within the handler to send a message with a random greeting to the specified recipient in the desired channel. Then, a new `SlackFunction` instance is declared that actually links the `helloWorld` handler to `SampleFunctionDefinition` through the function's ID:
```

```js
// Let's create a new Slack Function with helloWorld as its handler
const helloWorldFunc = new SlackFunction(SampleFunctionDefinition.id, helloWorld);
```

Once this file has been created, the function handler can be registered in an `index.js` file within the `listeners/functions` directory:
```js
// listeners/functions/index.js
const { helloWorldFunc } = require('./hello-world');

// Register a complete function
module.exports.register = (app) => {
  app.function(helloWorldFunc);
  // Register another function here
};
```

> You can see the full completed `hello-world.js` function listener in our [Bolt for JavaScript Starter Template](https://github.com/slack-samples/bolt-js-starter-template/blob/future/listeners/functions/hello-world.js).

Reviewer: ```suggestion

```
@stevengill - forgot to add this comment to my review - we can delete this part since we're linking to the `hello-world.js` file above!

---
title: Custom functions
order: 5
slug: custom-functions
lang: en
layout: tutorial
permalink: /future/custom-functions
---
# Custom functions <span class="label-beta">BETA</span>

<div class="section-content">
On the next-generation platform, you can build **custom functions**, reusable building blocks of automation that are deployed to our infrastructure and accept inputs, perform some calculations, and provide outputs. Functions can be used as steps in [Workflows](/bolt-js/future/workflows)&mdash;and Workflows are invoked by [Triggers](/bolt-js/future/triggers).

To create a function, we need to do the following: 
- [define the function](#define) in the Manifest;
- [implement the function](#implement) in its respective source file;
</div>

---

### Defining a function {#define}
Functions are defined in your app via the `DefineFunction` method, which is part of the SDK that gets included with every newly created project.

> Let's take a look at a simple function:

Author: ```suggestion
Let's go ahead and create a new function definition file under `manfiest/functions` directory. Name it something related to what the function does. In our [Bolt for JavaScript Starter Template](https://github.com/slack-samples/bolt-js-starter-template/blob/future/manifest/function/sample-function.js), we name the file `sample-function.js`. Let's take a peak at it:
```

```javascript
input_parameters: {
  properties: {
    customer_id: {
      type: Schema.types.string,
      description: "The customer's ID"
    }
  },
  required: ["customer_id"]
}
```

Functions can (and generally should) declare inputs and outputs. 

Inputs are declared in the `input_parameters` property, and outputs are declared in the `output_parameters` property. Each can contain either [Built-in types](https://api.slack.com/future/types) or [Custom types](https://api.slack.com/future/types/custom) that you define.

While, strictly speaking, input and output parameters are optional, they are a common and standard way to pass data between functions and nearly any function you write will expect at least one input and pass along an output.

Functions are similar in philosophy to Unix system commands: they should be minimalist, modular, and reusable. Expect the output of one function to eventually become the input of another, with no other frame of reference.

Once your function is defined in its own file in `manifest/function`, the next steps are to add a function listener, register the listener, and implement the function in a workflow.

---

> ### Implementing a function {#implement}

Reviewer 2: ```suggestion
### Implementing a Function {#implement}
```
Ooh ought these be capital? I don't really know...

Reviewer: i think it makes sense for them to be capital!!

Implement functions in just a few steps:

#### 1. Create the function file in `manifest/functions`
Create a file for your function to live in and name it something that makes sense for your function. For example, if I'm writing a function that sends one of our engineers Brad a singing telegram, I might name the file `send-brad-some-singers.js`.

> #### 2. Write code in the function file to define your function
> The exported module of your function's function file should be the function definition. It can be either asynchronous (for example, if you're calling API methods) or sychronous.
> 
> The function takes a single argument called its "context", and returns an object that exactly matches the structure of function definition's `output_parameters`.
> 
> Let's look at a sample function file, `sample-function.js`, from the `manifest/functions` directory in our default [Bolt for JavaScript Starter Template](https://github.com/slack-samples/bolt-js-starter-template/tree/future). This function, `SampleFunctionDefinition`, defines its expected input parameters:

Author: ```suggestion
```

---
title: Custom functions
order: 5
slug: custom-functions
lang: en
layout: tutorial
permalink: /future/custom-functions
---
# Custom functions <span class="label-beta">BETA</span>

<div class="section-content">
On the next-generation platform, you can build **custom functions**, reusable building blocks of automation that are deployed to our infrastructure and accept inputs, perform some calculations, and provide outputs. Functions can be used as steps in [Workflows](/bolt-js/future/workflows)&mdash;and Workflows are invoked by [Triggers](/bolt-js/future/triggers).

To create a function, we need to do the following: 
- [define the function](#define) in the Manifest;
- [implement the function](#implement) in its respective source file;
</div>

---

### Defining a function {#define}
Functions are defined in your app via the `DefineFunction` method, which is part of the SDK that gets included with every newly created project.

Let's take a look at a simple function:

> ```js

Reviewer: @srajiang a `source_file` property was mentioned here as "required" but i removed it because it seemed like we didn't need it. thoughts on adding it back in for the bolt context?

Reviewer 2: Yep source_file is only relevant for building and deploying to slack!
